### PR TITLE
UPDATED: Made is_last_serum_peak_group a MaintainedModel field

### PR DESCRIPTION
**Pull request by hepcat72**
## Summary Change Description

__I am creating an alternate PR because the merges/rebases up bring this up to date with main were a bit messy, so I did it on a separate branch to be able to easily re-try things when they went awry...__

Added the ability to propagate changes to children in MaintainedModel.  To do this, I converted 3 `@cached_function` methods to `@maintained_field_function` methods:

- `Animal._last_serum_sample()` updates new field `Animal.last_serum_sample`
- `Sample._is_serum_sample()` updates new field `Sample.is_serum_sample`
- `FCirc.is_last_serum_peak_group()` updates new field `FCirc.is_last`

I had to change the way `perform_buffered_updates` works to be able to propagate to both parents and children.  I also created a new decorator called `@maintained_model_relation` that is applied to classes that do not have maintained fields but need to be able to trigger updates of related models with maintained fields when they change.

And since the triggered updates need to happen when an M:M relation is updated with `.add()` (in addition to `.save()` and `.delete()` triggers), I added an `m2m_changed` handler, which I named: `m2m_propagation_handler`.

Lastly, I implemented the review issues from #493 in this branch.

## Affected Issue Numbers

- Partially addresses #460

## Code Review Notes

- Updated a few lingering example datasets. They happened to pass before because Animal.infusate was an optional field.  But based on a decision from the review of #493, that field is now required, which revealed the old formats.
- I had to create `create_is_null_field` in the model utilities that uses the ORM method called `.extra()` because I discovered that when you use `.order_by()`, it always puts null values at the end, which can thwart attempts to use `.last()`.  I used it when debugging the files that didn't have infusate columns, resulting in null values for the "last" serum sample and/or peak group even though one existed. It doesn't have much effect now that infusate is required, but will come in handy when infusate becomes optional.

## Checklist

- [ ] All issue requirements satisfied (or no linked issues)
- [x] [Linting passes](https://github.com/Princeton-LSI-ResearchComputing/tracebase/blob/main/CONTRIBUTING.md#linting).
  - The linting errors are fixed in main. I just need to rebase.
- [x] [Migrations created & committed *(or no model changes)*](https://github.com/Princeton-LSI-ResearchComputing/tracebase/blob/main/CONTRIBUTING.md#migration-process)
- Tests
  - [x] [Tests implemented *(or no code changes)*](https://github.com/Princeton-LSI-ResearchComputing/tracebase/blob/main/CONTRIBUTING.md#test-implementation)
  - [x] [All *multi_working* tag tests pass locally](https://github.com/Princeton-LSI-ResearchComputing/tracebase/blob/main/CONTRIBUTING.md#quality-control)
  - [x] All example load tests pass (remotely) *(or their failures predated and are unrelated to this branch)*
  - [x] `@tag("multi_working")` added to newly passing test functions/classes *(or no newly passing tests)*
  - [x] `@tag("multi_broken")`, `@tag("multi_unknown")`, or `@tag("multi_mixed")` removed from newly passing test functions/classes *(or no newly passing tests)*

### CODE REVIEW

File: DataRepo/migrations/0001_initial.py [hunk 1]
```diff
@@ -1,4 +1,4 @@
-# Generated by Django 3.2.5 on 2022-07-20 16:35
```
Comment (hepcat72) [inline DataRepo/migrations/0001_initial.py:1]: C. Rebuilt the migrations to address issues. The migration files may be ignored.

File: DataRepo/models/animal.py [hunk 2]
```diff
@@ -39,8 +44,8 @@ class Animal(HierCachedModel, ElementLabel):
         #       they are referenced through restricted foreign keys: 'Animal.infusate'.", {<Animal: 090320_M1>,
         #       <Animal: 090320_M2>, <Animal: 090320_M3>, ...
         on_delete=models.RESTRICT,
-        null=True,
-        blank=True,
+        null=False,
+        blank=False,
```
Comment (hepcat72) [inline DataRepo/models/animal.py:48]: C. This was a result of a decision made during review of https://github.com/Princeton-LSI-ResearchComputing/tracebase/pull/493

File: DataRepo/models/animal.py [hunk 3]
```diff
@@ -101,25 +106,40 @@ class Animal(HierCachedModel, ElementLabel):
...
+        related_name="animals",
+        help_text="Automatically maintained field. Shortcut to the last serum sample.",
+    )
 
     @property  # type: ignore
     @cached_function
     def tracers(self):
         if self.infusate.tracers.count() == 0:
-            warnings.warn(f"Animal [{self.animal}] has no tracers.")
+            warnings.warn(f"Animal [{self.name}] has no tracers.")
```
Comment (hepcat72) [inline DataRepo/models/animal.py:122]: C. This was an uncaught bug from the previous review of https://github.com/Princeton-LSI-ResearchComputing/tracebase/pull/493.

File: DataRepo/models/maintained_model.py [hunk 4]
```diff
@@ -151,68 +252,179 @@ def decorator(fn):
...
+
+                        m2m_changed.connect(
+                            m2m_propagation_handler,
+                            sender=through_model,
+                        )
+                    # m2m_changed.connect(toppings_changed, sender=Pizza.toppings.through)
+                except AttributeError as ae:
+                    if "has no attribute 'many_to_many'" not in str(ae):
+                        raise ae
+                    # Else - no propagation handler needed
```
Comment (hepcat72) [inline DataRepo/models/maintained_model.py:415]: C. This is the MaintainedModel validation code to ensure auto-updates are properly configured.

File: DataRepo/models/maintained_model.py [hunk 5]
```diff
@@ -226,20 +438,34 @@ def save(self, *args, **kwargs):
...
+        except (IntegrityError, ForeignKeyViolation) as uc:
+            if "violates foreign key constraint" in str(
+                uc
+            ) or "duplicate key value violates unique constraint" in str(uc):
+                warnings.warn(
+                    f"Ignoring {uc.__class__.__name__} exception during auto-update of {self.__class__.__name__}."
+                    f"{self.id}: [{str(uc)}]."
+                )
+            else:
+                raise uc
```
Comment (hepcat72) [inline DataRepo/models/maintained_model.py:453]: C. The errors which prompted me to add the try/catch no longer happen, but I decided to leave this change in because it will aid future development.

File: DataRepo/models/maintained_model.py [hunk 6]
```diff
@@ -250,48 +476,78 @@ def delete(self, *args, **kwargs):
...
                     current_val = getattr(self, update_fld)
-                except AttributeError:
-                    raise BadModelField(
-                        self.__class__.__name__, update_fld, update_fun.__qualname__
+                except Exception as e:
+                    warnings.warn(
+                        f"Unknown error getting current value: [{str(e)}].  Possibly due to this being triggered by a "
+                        "deleted record that is linked in a related model's maintained field."
                     )
+                    current_val = "<error>"
```
Comment (hepcat72) [inline DataRepo/models/maintained_model.py:501]: C. The errors which prompted me to add the try/catch no longer happen, but I decided to leave this change in because it will aid future development.

File: DataRepo/models/maintained_model.py [hunk 7]
```diff
@@ -151,68 +252,179 @@ def decorator(fn):
...
+        if parent_field_name is not None:
+            msg += f" trigger updates to parent: {class_name}." f"{parent_field_name}"
+        if parent_field_name is not None and len(child_field_names) > 0:
+            msg += " and "
+        if child_field_names is not None and len(child_field_names) > 0:
+            msg += (
+                f" trigger updates to children: "
+                f"{', '.join([class_name + '.' + c for c in child_field_names])}"
             )
+        print(f"{msg}.")
```
Comment (lparsons) [inline DataRepo/models/maintained_model.py:282]: NB: Perhaps add back the `if settings.DEBUG:`?

File: DataRepo/models/maintained_model.py [hunk 8]
```diff
@@ -151,68 +252,179 @@ def decorator(fn):
...
+
+    # TODO: Add controls to the call of obj.call_dfs_related_updaters() so that it is only called when auto_updates and
+    #       propagate are true.  See MaintainedModel.save().  In some cases, based on the value of reverse, the
+    #       "to model" object may need to be called instead.
+    # frm_mdl = obj.__class__
+    # to_mdl = kwargs.pop("model", None)
+    # frm_id = obj.pk
+    # rev = kwargs.pop("reverse", None)
+    # db = kwargs.pop("using", None)
+    # to_id = kwargs.pop("pk_set", None)
```
Comment (lparsons) [inline DataRepo/models/maintained_model.py:311]: Remove commented code?

File: DataRepo/models/maintained_model.py [hunk 9]
```diff
@@ -151,68 +252,179 @@ def decorator(fn):
...
+
+    # I do not need a "propagate" argument here like in save() and delete() to support not propagating during
+    # perform_buffered_updates because none of the maintained_field_functions currently use the .add() method and even
+    # if any did, I'm not sure triggering updates based on those would be necessary, because it's already propagating
+    # due to preform_buffered_updates having been called.  Regardless, it's not an issue.
```
Comment (lparsons) [inline DataRepo/models/maintained_model.py:316]: Slightly confusing comment.

**Review by jcmatese — APPROVED** [review]

File: DataRepo/models/animal.py [hunk 10]
```diff
@@ -101,25 +106,40 @@ class Animal(HierCachedModel, ElementLabel):
...
+        update_field_name="last_serum_sample",
+    )
+    def _last_serum_sample(self):
         """
         last_serum_sample in an instance method that returns the last single serum sample removed from the animal,
         based on the time elapsed/duration from the initiation of infusion or treatment, typically.  If the animal has
         no serum samples or if the retrieved serum sample has no annotated time_collected, a warning will be issued.
         """
+        # Create an is_null field for time_collected to be able to sort them
+        (extra_args, is_null_field) = create_is_null_field("time_collected")
```
[the remark this answers is not in the document]
Comment (hepcat72) [reply DataRepo/models/animal.py:138]: Does this explain it well enough or should I tweak the docstring?
```
def create_is_null_field(field_with_null):
    """
    The Django ORM's order_by always puts null fields at the end, which can be a problem if you want the last record
    ordered by a date field.  This method will return a dict that contains the `select` and `order_by arguments for the
    .extra(method, along with the name of the added "is null field that you can use as an argument to `order_by`)`.

    Example: extra_args, is_null_field = create_is_null_field("msrun__date")
    MSRun.objects.extra(**extra_args).order_by(f"-{is_null_field}", "msrun__date")

    Note, adding .annotate() doesn't seem to work, or rather I couldn't get it to work. See:
    https://stackoverflow.com/questions/7749216/django-order-by-date-but-have-none-at-end
    """
```

File: DataRepo/models/ms_run.py [hunk 11]
```diff
@@ -29,6 +38,8 @@ class MSRun(HierCachedModel):
     # Don't allow a Sample to be deleted if an MSRun links to it
     sample = models.ForeignKey(
         to="DataRepo.Sample",
+        # PR REVIEW NOTE: Seems that this should probably be CASCADE. If you delete a sample, you should delete its
+        #                 msruns...
```
[the remark this answers is not in the document]
Comment (hepcat72) [reply DataRepo/models/ms_run.py:42]: #511

File: DataRepo/models/peak_group.py [hunk 12]
```diff
@@ -3,9 +3,18 @@
...
+    maintained_model_relation,
+)
 
 
-class PeakGroup(HierCachedModel):
+@maintained_model_relation(
+    generation=3,
+    parent_field_name="msrun",
+    update_label="fcirc_calcs",
+)
```
[the remark this answers is not in the document]
Comment (hepcat72) [reply DataRepo/models/peak_group.py:16]: I think the decorator's docstring was copied/moved from a previous version and not sufficiently updated, so:

- [ ] Update the docstring for the `maintained_model_relation` decorator

but basically, you apply this decorator to a class that has no maintained fields, but you want changes to objects of this model to trigger updates of fields in other models.

Updates can either be triggered to parent models of child models, and to establish those relations (as in this case), you set the `parent_field_name` to the field in `PeakGroup` that is a foreign key to the parent record.

`generation` has been *partially* deprecated (but is still necessary). It had more utility when `perform_buffered_updates` executed updates in a BFS order (from leaves to root).  The generation number indicated the level of the tree the Model is in.  By adding the ability to go both directions in mass autoupdates, `perform_buffered_updates` now traverses the tree in DFS order and prevents repeated updates by keeping track of what's been updated and what hasn't.  Currently, `generation` is used primarily for validation that a decorator has the correct arguments (e.g. the root [generation=0] cannot have a `parent_field_name`).  It can also be used to limit what is updated, though with the current change, it's possible that may now be defunct.

File: DataRepo/models/peak_group_label.py [hunk 13]
```diff
@@ -275,13 +279,21 @@ def is_tracer_label_compound_group(self):
...
-        Instance method which returns True if a peakgroup was obtained from a
-        msrun of a serum sample. Uncertain whether this is a true concern.
+        Instance method which returns True if a peakgroup was obtained from a serum sample.
         """
-        if self.peak_group.msrun.sample.is_serum_sample:
+        if self.peak_group.msrun.sample.is_serum_sample is None:
+            warnings.warn(
+                f"Sample {self.peak_group.msrun.sample.name}'s is_serum_sample field hasn't been set."
+            )
+            fss = self.peak_group.msrun.sample._is_serum_sample()
```
[the remark this answers is not in the document]
Comment (hepcat72) [reply DataRepo/models/peak_group_label.py:288]: If `is_serum_sample` was a function and not a field, I would agree. This was just intended to make the code robust the failures to populate a maintained field.  In fact, I was intending to implement similar hardy code to see if I could make Fan's queries that rely on maintained fields to be more fault tolerant.

That said, I think that the ultimate solution here is (as is planned) to turn this method into a maintained field and always call `self.peak_group.msrun.sample._is_serum_sample()` in this function instead of relying on the population of a value in another maintained field, which I have documented as a best practive in the code of `MaintainedField`:

>    This is a decorator factory for functions in a Model class that are identified to be used to update a supplied
>     field and field of any linked parent/child record (for when the record is changed).  This function returns a
>     decorator that takes the decorated function.  **That function should not use the value of another maintained field in
>     its calculation because the order of update is not guaranteed to occur in a favorable series.**  It should return a
>     value compatible with the field type supplied.

File: DataRepo/models/sample.py [hunk 14]
```diff
@@ -60,19 +68,29 @@ class Sample(HierCachedModel):
...
+        child_field_names=["fcircs"],
+        update_field_name="is_serum_sample",
+        update_label="fcirc_calcs",
+    )
+    def _is_serum_sample(self):
         """returns True if the sample is flagged as a "serum" sample"""
-        return self.tissue.is_serum()
+        try:
+            iss = self.tissue.is_serum()
+        except Exception as e:
```
[the remark this answers is not in the document]
Comment (hepcat72) [reply DataRepo/models/sample.py:82]: Yeah, I was cognizant of that best practice when I wrote this.  I think that I'd added it initially as a "debug print" while at the time, searching for what was causing an exception.  I don't recall if I'd figured it out and fixed the source of the error, but I expect when it came to cleaning up the debug code, I no longer remembered which exception had caused the problem and decided to keep this code to make it robust/fault-tolerant, while providing info in the debug output to help me update the code later if I encountered it again.  In fact, I should probably add the class name to the output.

I'm amenable to either:

- Adding the class name to the output so the specific exception can be caught in the future
- Removing the try/catch entirely and just handling it anew when/if it happens again

Do you have a preference?

File: DataRepo/models/sample.py [hunk 15]
```diff
@@ -81,11 +99,14 @@ def last_tracer_peak_groups(self):
 
         # Get the last peakgroup for each tracer that has this label
         last_peakgroup_ids = []
+        (extra_args, is_null_field) = create_is_null_field("msrun__date")
+        print(f"Sample.py PeakGroup: Extra args: {extra_args}")
```
[the remark this answers is not in the document]
Comment (hepcat72) [reply DataRepo/models/sample.py:103]: Actually, I intended to remove all these and missed this one.

- [ ] TODO: Remove debug print

File: DataRepo/models/utilities.py [hunk 16]
```diff
@@ -144,3 +145,43 @@ def dereference_field(field_name, model_name):
 
 def get_model_by_name(model_name):
     return apps.get_model("DataRepo", model_name)
+
+
+def create_is_null_field(field_with_null):
```
[the remark this answers is not in the document]
Comment (hepcat72) [reply DataRepo/models/utilities.py:150]: I just saw code of yours where you had done that and it made me think of this.  I did consider this when implementing this solution...  Here's a counter-point: what if there's only 1 serum sample or only 1 instance of a tracer's peak group and either Sampletime_collected or the msrun.date is null.  Would we want to use that Sample or PeakGroup for the appearance/disappearance calculations?

File: DataRepo/tests/tracebase_test_case.py [hunk 17]
```diff
@@ -1,25 +1,71 @@
...
-    This wrapper of TestCase makes the necessary/desirable settings for all test classes.
+    Class creation factory where the base class is an argument.  Note, it must receive a TestCase-compatible class.
     """
 
-    maxDiff = None
-    databases = "__all__"
+    class TracebaseTestCaseTemplate(base_class):
+        """
+        This wrapper of both TestCase and TransactionTestCase makes the necessary/desirable settings for all test
+        classes and implements running time reporting.
```
[the remark this answers is not in the document]
Comment (hepcat72) [reply DataRepo/tests/tracebase_test_case.py:17]: Conceded.  Every time I do this, I feel guilty about giving Chris such a hard time about this sort of feature creep when working on TreeView.

File: DataRepo/utils/accucor_data_loader.py [hunk 18]
```diff
@@ -979,9 +978,39 @@ def parse_isotope_string(
...
+            raise e
+
+        # This cannot be in the atomic block because it needs to execute queries that generate the error:
+        # An error occurred in the current transaction. You can't execute queries until the end of the 'atomic' block.
+        # It comes from trying to trigger updates in many-related records, which uses `.count()` (to see if there exist
+        # records to propagate changes), `.first()` to see if the related model is a MaintainedModel (inside an
+        # isinstance call), and `.all()` to cycle through the related records
+
+        # TODO: This may no longer be necessary since various bugs have been fixed.  Put this back under
+        #       transaction.atomic and re-test
```
[the remark this answers is not in the document]
Comment (hepcat72) [reply DataRepo/utils/accucor_data_loader.py:1007]: Yeah, I wanted to test this before creating the PR, but I was trying to push things out fast.

And actually, the latest PR exhibits this same error after I re-added the call to `perform_buffered_updates` to the `sample_table_loader` code.  Weirdly enough, neither I nor Fan get the error that github reported and I'm looking into it now.  I will handle this in that PR.  I'll create a to-do comment there.

File: DataRepo/utils/sample_table_loader.py [hunk 19]
```diff
@@ -206,11 +208,48 @@ def load_sample_table(self, data, skip_researcher_check=False, debug=False):
...
+                if tracer_concs is None:
+                    raise NoConcentrations(
+                        f"{self.headers.INFUSATE} [{infusate_str}] supplied without "
+                        f"{self.headers.TRACER_CONCENTRATIONS}."
+                    )
+                infusate_data_object = parse_infusate_name(infusate_str, tracer_concs)
+                (infusate, created) = Infusate.objects.using(
+                    self.db
+                ).get_or_create_infusate(infusate_data_object)
+
```
[the remark this answers is not in the document]
Comment (hepcat72) [reply DataRepo/utils/sample_table_loader.py:245]: Yeah, it was just moved.  It previously only executed if an Animal record was created.  That was a hold-over from when Animal contained the tracer fields.  Now that Infusate (and related child models) are in a separate table, it had to be created independent of the creation of an Animal record.

File: DataRepo/tests/models/test_sample.py [hunk 20]
```diff
@@ -0,0 +1,73 @@
...
+    def test_is_serum_sample_autoupdates(self):
+        """
+        Issue #460, test 3.1.2.
+        3. Updates of FCirc.is_last, Sample.is_serum_sample, and Animal.last_serum_sample are triggered by themselves
+           and by changes to models down to PeakGroup.
+          1. Create a new serum sample whose time collected is later than existing serum samples.
+            2. Confirm the new sample's Sample.is_serum_sample is True.
+        """
+        # Assert that the new serum sample's is_serum_sample is autoupdated to True
+        self.assertTrue(self.newlss.is_serum_sample)
```
[the remark this answers is not in the document]
Comment (hepcat72) [reply DataRepo/tests/models/test_sample.py:73]: That's a good point.

- [ ] TODO: Associate the test description with the code that implements it.

File: DataRepo/tests/models/test_sample.py [hunk 21]
```diff
@@ -0,0 +1,73 @@
...
+        # Now create a new last serum sample (without any peak groups)
+        tissue = lss.tissue
+        tc = lss.time_collected + timedelta(seconds=1)
+        nlss = Sample.objects.create(animal=animal, tissue=tissue, time_collected=tc)
+
+        self.lss = lss
+        self.newlss = nlss
+
+    @classmethod
+    def setUpTestData(cls):
```
[the remark this answers is not in the document]
Comment (hepcat72) [reply DataRepo/tests/models/test_sample.py:35]: Yeah, I'm still getting the hang of how the data setup works. I think I'll just move the code currently in setUp into the test function.

File: DataRepo/models/maintained_model.py [hunk 22]
```diff
@@ -105,34 +139,101 @@ def updater_list_has_labels(updaters_list, label_filters):
...
-def field_updater_function(
-    generation, update_field_name=None, parent_field_name=None, update_label=None
+def maintained_model_relation(
+    generation, parent_field_name=None, child_field_names=[], update_label=None
+):
+    """
+    Use this decorator to add connections between classes when it does not have any maintained fields.  For example,
+    if you only want to maintain 1 field in 1 class, but you want changes in a related class to trigger updates to
+    that field, call this method in the related class's __init__ method in order to trigger those updates of the
+    field in the other class.
```
[the remark this answers is not in the document]
Comment (hepcat72) [reply DataRepo/models/maintained_model.py:149]: Yeah, this is the one I indicated above needs to be updated.  Providing an example usage is a good suggestion.

File: DataRepo/models/maintained_model.py (hunk 7 above)
Comment (hepcat72) [reply DataRepo/models/maintained_model.py:282]: OK.  I wasn't getting it when I needed/expected it, which was why I removed that conditional.  At the time, I did wonder why settings.DEBUG didn't seem to be working.  Perhaps it had to do with overrides above tests.  I just didn't look into it.

- [ ] Re-add settings.DEBUG conditionals to the decorator addition messages

File: DataRepo/models/maintained_model.py [hunk 23]
```diff
@@ -151,68 +252,179 @@ def decorator(fn):
...
+    fields, an m2m_changed signal is attached to all M:M fields in MaintainedModel.__init__ to tell us when a
+    MaintainedModel has an M:M field that has been added to.  That causes this method to be called, and from here we
+    can propagate the changes.
+    """
+    obj = kwargs.pop("instance", None)
+    act = kwargs.pop("action", None)
+
+    # TODO: Add controls to the call of obj.call_dfs_related_updaters() so that it is only called when auto_updates and
+    #       propagate are true.  See MaintainedModel.save().  In some cases, based on the value of reverse, the
+    #       "to model" object may need to be called instead.
```
[the remark this answers is not in the document]
Comment (hepcat72) [reply DataRepo/models/maintained_model.py:305]: #512

File: DataRepo/models/maintained_model.py (hunk 8 above)
Comment (hepcat72) [reply DataRepo/models/maintained_model.py:311]: - [ ] Remove commented code saved in issue #512

File: DataRepo/models/maintained_model.py (hunk 9 above)
Comment (hepcat72) [reply DataRepo/models/maintained_model.py:316]: Yeah, I'm not sure I fully understand the implications, which is probably why it's confusing.  I have to think about it.

File: DataRepo/models/maintained_model.py [hunk 24]
```diff
@@ -151,68 +252,179 @@ def decorator(fn):
...
-            parent_msg = ""
-            if parent_field_name is not None:
-                parent_msg = (
-                    f" trigger updates of maintained fields in model reference by foreign key: {class_name}."
-                    f"{parent_field_name}"
-                )
-            print(
-                f"Added field_updater_function decorator to function {fn.__qualname__} in order to{local_msg}"
-                f"{parent_msg}."
+        # if settings.DEBUG:
```
[the remark this answers is not in the document]
Comment (hepcat72) [reply DataRepo/models/maintained_model.py:267]: Lance suggested I re-add it.

File: DataRepo/models/fcirc.py [hunk 25]
```diff
@@ -33,6 +43,13 @@ class FCirc(HierCachedModel):
         default=ElementLabel.CARBON,
         help_text='An element that is labeled in any of the tracers in this infusate (e.g. "C", "H", "O").',
     )
+    is_last = models.BooleanField(
+        default=False,
+        help_text=(
+            "This field indicates whether the last peak group of this serum sample and this tracer, is the last among "
+            "the serum samples/tracers for the associated animal. Maintained field. Do not edit/set."
+        ),
+    )
```
[the remark this answers is not in the document]
Comment (hepcat72) [reply DataRepo/models/fcirc.py:53]: There already exists code that raises an exception if this is attempted:

https://github.com/Princeton-LSI-ResearchComputing/tracebase/blob/248242fbbc157eaa253e984b308f98291d3d6376/DataRepo/models/maintained_model.py#L345-L348

The help text is to try and head such efforts off at the pass.

File: DataRepo/models/utilities.py (hunk 16 above)
[the remark this answers is not in the document]
Comment (lparsons) [reply DataRepo/models/utilities.py:150]: I could see handling that case either way. TBH, I'm not sure we need to know the "last" sample at all. It seems better to simply calculate Ra/Rd for any serum sample and let the researcher make the interpretation of those values.

[the remark this answers is not in the document]
Comment (hepcat72) [reply DataRepo/models/utilities.py:150]: We do calculate for every one. Your question makes me realize that my query was imprecise, so let me rephrase. If the user is filtering the FCirc output for only "last serum tracer peak groups" and there's only one such peak group, but one or both of those date fields is null, should we include it as a "last serum tracer peak group" or not?

[the remark this answers is not in the document]
Comment (lparsons) [reply DataRepo/models/utilities.py:150]: Yeah, I realize that is how it works now. The fact that answering that question isn't straightforward is why I would lean toward removing the "last serum tracer peak group" filter altogether. As long as the "time collected" is reported, let the researcher interpret things. There may be data not uploaded, or the time might be too short to consider things "stable" even if it is the "last" sample collected.

[the remark this answers is not in the document]
Comment (hepcat72) [reply DataRepo/models/utilities.py:150]: `time_collected` can definitely be one of the search parameters, though are you saying that there's no value in differentiating between the last serum tracer peak group and any other? That requirement, established quite some time back, has resulted in significant effort. I feel like it would be a shame to throw it out and change the requirements when we're mostly done with the effort to get there.  Perhaps we should revisit the motivations behind having that in the first place.  And what is the impact to the researcher if it's removed?  I selected these fields to work on initially WRT #460 because of the subsequent issue: #500, which has a dependency on this issue.

File: DataRepo/tests/tracebase_test_case.py [hunk 26]
```diff
@@ -1,25 +1,71 @@
...
+def reportRunTime(id, startTime):
     """
-    This wrapper of TestCase makes the necessary/desirable settings for all test classes.
+    Print the runtime of a test given the test ID and start time.
     """
 
-    maxDiff = None
-    databases = "__all__"
+    # TODO: When issue #480 is implemented, a test should be added here to ensure each test runs in under some
+    #       reasonable threshold.  There might even be a way to ensure the data setup runs quickly as well.
```
[the remark this answers is not in the document]
Comment (hepcat72) [reply DataRepo/tests/tracebase_test_case.py:57]: Such as issue #480? Or are you referring to something distinctly different from that issue?

[the remark this answers is not in the document]
Comment (lparsons) [reply DataRepo/tests/tracebase_test_case.py:57]: I didn't see anything about setting a maximum run time for tests is #480. I'm just thinking that suggestion should go in a relevant issue and not as a comment in the code.

File: DataRepo/models/utilities.py (hunk 16 above)
[the remark this answers is not in the document]
Comment (lparsons) [reply DataRepo/models/utilities.py:150]: We don't have to remove it right now or even change anything at the moment. But I do think that complexities like this illustrate that we should be careful to understand the implications of the choices we are making. I worry researchers will just assume we did "the right thing" and interpret whatever values flag as "last" as Ra/Rd values suitable for approximating "circulatory flux".

**Comment by hepcat72** [comment]
> The changes to the loading code were surprising, since ideally this PR would be on top of code that was already functional.

The changes have pretty straightforward reasons:

1. tracer information is now in a related model, and was made required, so to create the Animal, an infusate must be supplied to the animal constructor, so it was moved above the animal constructor
2. New functionality of MaintainedModel caused TransactionManagementErrors, so perform_buffered_updates had to be moved outside of the atomic transaction block.

> Ideally, we could create some issues for identified bugs (loading code, etc.), new features (test timing), changes required to MaintainedModel to support new functionality, and then changes to the models to implement that.

Implementing the MaintainedModel updates required model changes to guide development.  I wouldn't have known what to anticipate without fields to develop for.

> This would allow us to spread the work accross multiple developers and have shorter coding and review cycles. It also helps to ensure we have working branches so others can work on other unrelated functionality such as updating of views, importing data, etc.

I accept the PR size criticism.  I definitely could make progress in curtailing my propensity for feature creep, but I think that the only real benefit in this case is the difficulty of the review.  I think the hold up WRT views and importing was more due to of a lack of communication than a development impediment.  There was nothing to prevent progress in either of those fronts, except for a mistaken push to main that was not the result of the scope of this PR but rather human error.  WRT views, I should have communicated that that code should be able to handle faults in the population of maintained fields.

File: DataRepo/tests/tracebase_test_case.py (hunk 26 above)
[the remark this answers is not in the document]
Comment (hepcat72) [reply DataRepo/tests/tracebase_test_case.py:57]: OK, well, I suggest that that is the issue where the discussion can occur.  I respect the desire to remove the comment, but I feel it's a good placeholder/reminder.  If I were to remove it and create a permalink to that spot, it would be tied to that commit.  Whereas if we keep the comment, it will keep up with code changes.  However, I think it's stable enough to just reference the method name in the issue, so:

- [ ] Remove the "issue #480" comment and referenc ethe method name in the issue

**Comment by lparsons** [comment]
Thanks for explaining the loading changes, makes sense. As for communication, that can be hard. I guess this is just a reminder that we can do better sharing what we're working and what priorities are at our dev meetings on Monday. Great work on this (and the other) PRs. I'm very excited about the progress on TraceBase!

File: DataRepo/models/utilities.py (hunk 16 above)
[the remark this answers is not in the document]
Comment (hepcat72) [reply DataRepo/models/utilities.py:150]: I do think that that's a valuable point and I totally agree, so I think I should:

- [ ] Create a warning-style user-facing indication that a "last serum tracer peak group" is based on either a sample without a time_collected or an msrun without a date

[the remark this answers is not in the document]
Comment (hepcat72) [reply DataRepo/models/utilities.py:150]: Created #514

**Comment by hepcat72** [comment]
Review items were implemented in PR #510
